applications/ParticleMechanicsApplication/custom_strategies/schemes/mpm_explicit_scheme.hpp

```diff
@@ -0,0 +1,565 @@
+
+                    #pragma omp parallel for
+                    for (int i = 0; i < static_cast<int>(rElements.size()); ++i)
+                    {
+                        auto it_elem = it_elem_begin + i;
+                        std::vector<bool> dummy;
+                        it_elem->CalculateOnIntegrationPoints(CALCULATE_EXPLICIT_MP_STRESS, dummy, rCurrentProcessInfo);
+                    }
+                }                
+
+                // Finalizes solution step for all of the conditions
+                const auto it_cond_begin = rModelPart.ConditionsBegin();
+                #pragma omp parallel for
+                for (int i = 0; i < static_cast<int>(rModelPart.Conditions().size()); ++i) {
+                    auto it_cond = it_cond_begin + i;
+                    it_cond->FinalizeSolutionStep(rCurrentProcessInfo);
+                }
+
+                KRATOS_CATCH("")
+            }
+            //***************************************************************************
+            //***************************************************************************
+
+            void CalculateUpdatedGridVelocityField(const ProcessInfo& rCurrentProcessInfo, ModelPart& rModelPart)
+            {
+                if (rCurrentProcessInfo.GetValue(EXPLICIT_STRESS_UPDATE_OPTION) == 1)
+                {
+                    // USL
+                    const SizeType DomainSize = rCurrentProcessInfo[DOMAIN_SIZE];
+                    #pragma omp parallel for
+                    for (int iter = 0; iter < static_cast<int>(rModelPart.Nodes().size()); ++iter)
+                    {
+                        NodeIterator i = rModelPart.NodesBegin() + iter;
+                        if ((i)->Is(ACTIVE))
+                        {
+                            array_1d<double, 3 >& r_nodal_momenta = (i)->FastGetSolutionStepValue(NODAL_MOMENTUM);
+                            array_1d<double, 3>& r_current_velocity = i->FastGetSolutionStepValue(VELOCITY);
+                            r_current_velocity.clear();
+                            const double nodal_mass = i->FastGetSolutionStepValue(NODAL_MASS);
+                            if (nodal_mass > numerical_limit)
+                            {
+                                for (IndexType j = 0; j < DomainSize; j++)
+                                {
+                                    r_current_velocity[j] = r_nodal_momenta[j] / nodal_mass;
+                                } // for DomainSize
+                            }
+                        }
+                    }
+                }
+                else if (rCurrentProcessInfo.GetValue(EXPLICIT_STRESS_UPDATE_OPTION) == 2)
+                {
+                    // MUSL stress update. This works by projecting the updated particle
+                    // velocity back to the nodes. The nodal velocity field is then
+                    // used for stress computations.
+
+                    // Reset grid velocities
```

Comment:
I thought there was a function for this in the variable utils

Reply:
There's a function in the explicit utils - but this is on the 'element' level. It's called in this part, after the grid velocity field is completely reset.